rust/hexane/src/slab/writer.rs

```diff
@@ -0,0 +1,531 @@
+            self.actions.get(self.slab_head),
+            Some(&WriteAction::SlabHead)
+        );
+        self.actions[self.slab_head] =
+            WriteAction::Slab(self.items, self.acc, self.abs, self.width);
+        self.num_slabs += 1;
+        self.slab_head = self.actions.len();
+        self.actions.push(WriteAction::SlabHead);
+    }
+
+    fn check_copy_overflow(&mut self, copy: usize) {
+        if self.width + copy > self.max && self.width > 0 && !self.locked {
+            self.close_lit();
+            self.close_slab();
+            self.width = 0;
+            self.acc = Acc::new();
+            self.bools = 0;
+            self.items = 0;
+        }
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.actions.len() <= 1 // first action is slab_head
+    }
+
+    pub fn write(mut self, out: &mut Vec<u8>) {
+        self.close_lit();
+        for action in self.actions {
+            action.write(out)
+        }
+    }
+
+    pub fn write_and_remap<'b, F>(mut self, out: &mut Vec<u8>, f: F)
+    where
+        F: Fn(&P) -> Option<&'b P>,
+        P: 'b,
+    {
+        self.close_lit();
+        for action in self.actions {
+            action.write_and_remap(out, &f)
+        }
+    }
+
+    pub fn into_column<C: ColumnCursor>(self, length: usize) -> ColumnData<C> {
+        let mut slabs = self.finish();
+        C::compute_min_max(&mut slabs);
+        ColumnData::init(length, SlabTree::load(slabs))
+    }
+
+    pub fn finish(mut self) -> Vec<Slab> {
+        self.close_lit();
+        if self.items > 0 {
+            self.close_slab();
+        }
+        if self.num_slabs == 0 {
+            return vec![];
+        }
+        self.actions.pop();
+        let mut result = Vec::with_capacity(self.num_slabs);
+        let mut buffer = vec![];
+        let mut len = 0;
+        let mut acc = Acc::new();
+        let mut abs = self.init_abs;
+        let mut next_abs = 0;
+        let mut width = 0;
+        for action in self.actions {
+            match action {
+                WriteAction::Slab(next_len, next_acc, next_next_abs, next_width) => {
+                    if !buffer.is_empty() {
+                        debug_assert_eq!(width, buffer.len());
+                        let data = std::mem::take(&mut buffer);
+                        result.push(Slab::new(data, len, acc, abs));
+                        abs = next_abs;
+                    }
+                    buffer = Vec::with_capacity(next_width);
+                    acc = next_acc;
+                    len = next_len;
+                    width = next_width;
+                    next_abs = next_next_abs;
+                }
+                action => {
+                    action.write(&mut buffer);
+                }
+            }
+        }
+        debug_assert_eq!(width, buffer.len());
+        result.push(Slab::new(buffer, len, acc, abs));
+        assert_eq!(self.num_slabs, result.len());
+        result
+    }
+
+    pub fn copy(
```

Comment:
I think we should add a struct for these arguments, otherwise it will be very easy to confuse the `lit` and `size` arguments (I have confused them several times whilst reading this code base).

Something like:

```
struct CopyArgs{
    slab: &'a [u8],
    range: Range<usize>,
    lit_length: usize,
    total_items: usize,
    acc: Acc,
    bool_state: Option<bool>,
  }
  ```
  
  Actuall `bool_state` needs another name, I've already forgotten what it indicates.